### PR TITLE
Throw error for duplicate component names in manifests YML

### LINKED ISSUE #3694: Detecting and alerting of duplication keys/components/entries in YAML file
### Is your feature request related to a problem? Please describe

it was found in release 1.3.11 , a PR to update [manifest](https://github.com/opensearch-project/opensearch-build/blob/main/manifests/1.3.11/opensearch-1.3.11.yml) has duplicated components name.
It would cause the resource wasted on CI to rebuild the duplicated components 

### Describe the solution you'd like

We want to have a check to detect if there is any duplication entries based on keys/components/names and probably fail the GitHub check

### Describe alternatives you've considered

Manually check for duplicate values

### Acceptance Criteria
* The manifest check should fail at CI level for components with duplicate components.name values in opensearch and opensearch-dashboard as well as test manifests. See what are [manifests](https://github.com/opensearch-project/opensearch-build/wiki/Building-an-OpenSearch-and-OpenSearch-Dashboards-Distribution#what-are-manifests)

### PR DESCRIPTION
**Pull request by rblcoder**
### Description
Add run yamllint in manifests github workflow

### Issues Resolved
 closes #3694

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by rblcoder** [comment]
Hi @gaiksaya , I am wondering, since the yaml-lint.yml is already checking all files, would it not also catch duplicate keys when raising a PR?
https://yamllint.readthedocs.io/en/stable/rules.html#module-yamllint.rules.key_duplicates

**Comment by gaiksaya** [comment]
> Hi @gaiksaya , I am wondering, since the yaml-lint.yml is already checking all files, would it not also catch duplicate keys when raising a PR? https://yamllint.readthedocs.io/en/stable/rules.html#module-yamllint.rules.key_duplicates

Thats the thing it does not. I believe it only checks the at first level but not at nested one. Here are the lint rules https://github.com/opensearch-project/opensearch-build/blob/main/.yamllint.yml

**Comment by rblcoder** [comment]
Which is the manifest file (with the errors) for which yamllint was unable to raise errors?

**Comment by gaiksaya** [comment]
> @gaiksaya Which is the manifest file (with the errors) for which yamllint was unable to raise errors?

Yes!

**Comment by rblcoder** [comment]
@gaiksaya I have made changes, could you please review them?

File: src/ci_workflow/ci_manifests.py [hunk 1]
```diff
@@ -23,6 +26,19 @@ def __klass(filename: str) -> Union[Type[CiTestManifest], Type[CiInputManifest]]
...
+    @staticmethod
+    def __check_duplicate_component_names(file: TextIOWrapper) -> None:
+
+        yaml_dict = yaml.safe_load(file)
+        component_names = []
+        for component in yaml_dict['components']:
+            component_names.append(component['name'])
+
+        if set(Counter(component_names).values()) != set([1]):
+            raise ValueError(f"Duplicate components in manifest {file.name}")
```
Comment (gaiksaya) [inline src/ci_workflow/ci_manifests.py:38]: Is it possible to display which component is duplicated? Also can you add respective unit test cases for this method?

Comment (rblcoder) [reply src/ci_workflow/ci_manifests.py]: I have added display of duplicate component names and unit tests, please have a look.

File: src/ci_workflow/ci_manifests.py [hunk 2]
```diff
@@ -23,6 +26,31 @@ def __klass(filename: str) -> Union[Type[CiTestManifest], Type[CiInputManifest]]
...
+        for component in yaml_dict['components']:
+            component_names.append(component['name'])
+        count_component_names = Counter(component_names)
+
+        if set(count_component_names.values()) != set([1]):
+            duplicate_component_names = CiManifests.__get_duplicate_component_names(count_component_names)
+            duplicate_component_names_string = ', '.join(duplicate_component_names)
+            raise ValueError(f"Duplicate components in manifest {file.name}. "
+                             f"Following are the components:"
+                             f"\n{duplicate_component_names_string}")
```
Comment (gaiksaya) [inline src/ci_workflow/ci_manifests.py:50]:
```suggestion
            raise ValueError(f"Found {duplicate_component_names_string} as a duplicate component(s) in manifest {file.name}. "
```

**Comment by gaiksaya** [comment]
Can you fix the tests @rblcoder ?
Thanks!

**Comment by gaiksaya** [comment]
File name typo:
`tests/tests_ci_workflow/test_duplcate_components_in_manifest_file.py` - > `tests/tests_ci_workflow/test_duplicate_components_in_manifest_file.py`

**Comment by rblcoder** [comment]
@gaiksaya I have made the change in the test filename.

**Review by gaiksaya — APPROVED** [review]

File: src/ci_workflow/ci_manifests.py [hunk 3]
```diff
@@ -23,6 +26,29 @@ def __klass(filename: str) -> Union[Type[CiTestManifest], Type[CiInputManifest]]
...
+        for component in yaml_dict['components']:
+            component_names.append(component['name'])
+        count_component_names = Counter(component_names)
+
+        if set(count_component_names.values()) != set([1]):
+            duplicate_component_names = CiManifests.__get_duplicate_component_names(count_component_names)
+            duplicate_component_names_string = ', '.join(duplicate_component_names)
+            raise ValueError(f"Found {duplicate_component_names_string} as a duplicate component(s) in manifest {file.name}. ")
+        file.seek(0)
+
```
Comment (dblock) [inline src/ci_workflow/ci_manifests.py:50]: I'm late to this PR, but this implementation seems unnecessarily complicated. 

I would load the items, then check with a 1-liner from https://stackoverflow.com/questions/9835762/how-do-i-find-the-duplicates-in-a-list-and-create-another-list-with-them.

File: tests/tests_ci_workflow/test_duplicate_components_in_manifest_file.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import unittest
+
+from ci_workflow.ci_manifests import CiManifests
+
+
+class TestDuplicateComponentsInManifestFile(unittest.TestCase):
+    def test_duplicate_components_in_manifest_file(self) -> None:
+        data_path = os.path.join(os.path.dirname(__file__), "data")
+        manifest_filename = os.path.join(data_path, 'opensearch-1.3.11.yml')
+        with self.assertRaises(ValueError):
```
Comment (dblock) [inline tests/tests_ci_workflow/test_duplicate_components_in_manifest_file.py:18]: You should be checking the actual error, as many errors could be `ValueError`.

I would also rename opensearch-1.3.11.yml to opensearch-1.3.11-with-duplicates.yml.

**Pull request merged**